### ISSUE TITLE
Performace: Sparse Arrays vs Dense Arrays

### ISSUE THREAD

**Issue by ilteroi** [LoneGazebo/Community-Patch-DLL#9181]
Right now there are many places where an iteration of all resource types or all unit types or all building types is needed to find non-zero database entries. Mostly there is only a single non-zero entry, making this approach inefficient in both space and time.

Storing only non-zero entries in a vector would be better in many cases. Profiling data is needed.

**Comment by LoneGazebo** [comment]
Yeah but damn, arrays don’t like to be resized.

**Comment by LoneGazebo** [comment]
Also so much of the UI and XML relies on array numbers matching uniformly, would you even be able to parse that without totally rewriting the data backend?

**Comment by ilteroi** [comment]
well that's where the profiling comes in ... doesn't need to be changed across the board.

the resizing part is not a problem in any case, the database entries don't change mid-game.

**Comment by IanE9** [comment]
Assuming you're talking about there being gaps between IDs leading to `NULL` pointers being present in various array like structures, fixing that soundly could probably be achieved by just updating the ID columns of every table that matters during `CvDllDatabaseUtility::PerformDatabasePostProcessing`. It may be possible to do this quickly by repeatedly taking the row with the highest ID and updating it to the lowest missing ID until there are no more gaps.

**Comment by ilteroi** [comment]
no, this is about database columns which are mostly empty. so for example we have to iterate 214 promotions to find whether a given unit gets it by default (bool pkUnitEntry->GetFreePromotions) or over 148 unit types to find out whether a building unlocks one of them (bool pkUnitInfo->GetBuildingClassRequireds)

would be much better to have CvUnitEntry::GetFreePromotions() returning a const vector<PromotionTypes>& or CvUnitEntry::GetPrereqBuildingClasses() returning a const vector<BuildingClassTypes>&

**Comment by LoneGazebo** [comment]
Isn’t that how civ traits are managed? Or perhaps beliefs? It’s been awhile since I looked at it last.

**Comment by ilteroi** [comment]
exception confirm the rule ... also i made some point fixes in the worst places already but it's just so stupid in general. just wanted to put the topic on the official todo list.
